### ISSUE TITLE
Build ignores local directory

### ISSUE THREAD

**Issue by evanshortiss** [openshift/source-to-image#904]
Hey, I've come across an unusual issue. I was trying to build my local repository in which a new *dev* branch was checked out using the following command:

```
$ git checkout -b dev
# make some code changes...now build
$ s2i build . registry.access.redhat.com/rhscl/nodejs-8-rhel7 openshift-ts
```

It continued to build my old *master* branch code. I checked the help output from *s2i build --help* and noted that a *--ref* flag was supported and this tipped me off that s2i might be having some issue with the Git branch/ref. 

I pushed my newly created local *dev* branch to my remote and did a build using:

```
$ s2i build --r dev . registry.access.redhat.com/rhscl/nodejs-8-rhel7 openshift-ts
```
This seemed to resolve the issue. For subsequent local *dev* builds I can now simply use the following (without the *-r* flag) and it works as I would expect:

```
$ s2i build . registry.access.redhat.com/rhscl/nodejs-8-rhel7 openshift-ts
```

This seems like a bug, right?

**Comment by evanshortiss** [comment]
I have not confirmed, but here are the steps I performed to discover this bug.

1. Git clone a repository with just a single branch; *master*
2. Run `git checkout -b dev` inside the cloned repo
3. Run `s2i build <source> <image> [<tag>]`

This should result in the *master* branch being built, but we expect that the locally checked out *dev* branch would.

**Comment by evanshortiss** [comment]
After a little more testing it seems that only code I've pushed builds. Perhaps this should be highlighted a little better? The `# Build from a local directory` in the `s2i build .` help output lead me to believe it was using my current checked out code.

**Comment by bparees** [comment]
it should be building the currently committed code.

Effectively we just do a "git clone . /tmp/target"  (followed by a checkout if you specified a ref).

If you want to avoid the git behavior entirely, use "--copy" which just does a filesystem copy instead of git manipulation.  The original design(perhaps flawed in hindsight) assumed you'd be referencing remote git repos as the input normally.

When i created a new local branch, committed changes to it, and then ran "s2i build ." on that dir (w/ the branch still checked out), it appears to use the content from the local branch as i would expect, do you still think you're seeing an issue with that behavior?

**Comment by evanshortiss** [comment]
> it should be building the currently committed code.

@bparees that's the conclusion I've come to here, thanks for the clarification. 

What tripped me up was when I first used source to image a while back I was testing it using a local codebase that was not a Git repo. In that instance `s2i build . <image> [<tag>]` always behaves as though it has been passed `--copy`.

Based on that previous experience I assumed that `s2i build . <image> [<tag>]` would always build the current state of the local directory per the help output (and since I was giving it a path, not a Git ref). Essentially, I expected it to do what `--copy` does, but it turns out that it behaves differently based on whether the repo is initialised with Git or not.

It doesn't feel intuitive to pass s2i a path, but then have it ignore the state of that path and instead use a commit found at that path - that's just my 2 cents though! The copy flag covers it, so perhaps changing `Build from a local directory` to `Build from a local directory. If this directory is using Git then the current commit will be built` might be clearer?

Mostly a case of RTFM vs. intuition.

**Comment by bparees** [comment]
> back I was testing it using a local codebase that was not a Git repo. In that instance s2i build . <image> [<tag>] always behaves as though it has been passed --copy.

yes that's our magic "default to copy instead of clone if the dir is not a git repo" behavior. :/

> but it turns out that it behaves differently based on whether the repo is initialised with Git or not.

yup

> It doesn't feel intuitive to pass s2i a path, but then have it ignore the state of that path and instead use a commit found at that path - that's just my 2 cents though!

no disagreement and you're not the first to run into this, however we're very reluctant to change default behaviors since it could break people's scripts/expectations at this point.

> The copy flag covers it, so perhaps changing Build from a local directory to Build from a local directory. If this directory is using Git then the current commit will be built might be clearer?

Good suggestion.  https://github.com/openshift/source-to-image/pull/906
